type_classes.md

```diff
@@ -583,20 +857,40 @@ class inductive Decidable (p : Prop) where
 # end Hidden
 ```
 
+BUGBUG: without reading inductive types I have no idea how to read this.
+After reading inductive types I still have no idea how to read this...
+
 Logically speaking, having an element ``t : Decidable p`` is stronger
 than having an element ``t : p ∨ ¬p``; it enables us to define values
 of an arbitrary type depending on the truth value of ``p``. For
 example, for the expression ``if p then a else b`` to make sense, we
 need to know that ``p`` is decidable. That expression is syntactic
 sugar for ``ite p a b``, where ``ite`` is defined as follows:
 
+BUGBUG: no idea what that first sentence means.  How could anything be
+stronger than ``t : p ∨ ¬p`` which is provably true?  What does stronger
```

Comment:
`p ∨ ¬p` for arbitrary `p : Prop` is in fact not provably true in Lean per default, you have to pull in the so called axiom of the excluded middle from classical logic via `open Classical` if you want to prove this.

This is because default (aka no excluded middle) Lean requires you to construct a proof of either `p` or or `¬p` in order to prove `p ∨ ¬p`, a well known example where this does not work would be proving that any turing machine `T` either halts or does not halt, you cannot always construct evidence for `halt T` or `¬(halt T)` since the problem is not decidable (hence this would be something that does not implement the `Decidable` typeclass).

While this does seem confusing from a classical perspective (of course everything is true or false right?) it does make sense from the Lean perspective where `p ∨ ¬p` does not only mean p is either true or false but also that for any `p` we can construct evidence that it is either true or false which is not the case.

If a proposition is stronger than another one this means that it has more guarantees (and thus possibilities) for us than the other one. A `Decidable` instance for a specific `p` is in this sense stronger than the axiom of the excluded middle that for all `p` `p ∨ ¬p`. This is because due to the pattern matching on `Decidable` we can actually write functions that do different things based on whether its `p` or `¬p` that is true, the general statement of the axiom of the excluded middle does not allow this since it does not construct a proof of `p` or `¬p` but merely states that either of them are true without proof.